pkg/provider/instances_v2.go

```diff
@@ -0,0 +1,193 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	cloudprovider "k8s.io/cloud-provider"
+	v1helper "k8s.io/cloud-provider/node/helpers"
+	"k8s.io/klog/v2"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// instanceIDLabelKey could be injected by k8s providers to find a corresponding virtual machine instance
+	// the value should be a virtual machine name
+	instanceIDLabelKey = "node.kubernetes.io/instance-id"
+)
+
+// Must match providerIDs built by cloudprovider.GetInstanceProviderID
+var providerIDRegexp = regexp.MustCompile(`^` + ProviderName + `://([0-9A-Za-z_-]+)$`)
+
+type instancesV2 struct {
+	namespace string
+	client    client.Client
+	config    InstancesV2Config
+}
+
+// InstanceExists returns true if the instance for the given node exists according to the cloud provider.
+func (i *instancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
+	instanceID, err := instanceIDFromProviderID(node.Spec.ProviderID)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = InstanceByVMIName(instanceID).Get(ctx, i.client, i.namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("Unable to find virtual machine instance %s", instanceID)
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+func (i *instancesV2) InstanceShutdown(ctx context.Context, node *corev1.Node) (bool, error) {
+	instanceID, err := instanceIDFromProviderID(node.Spec.ProviderID)
+	if err != nil {
+		return false, err
+	}
+
+	instance, err := InstanceByVMIName(instanceID).Get(ctx, i.client, i.namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	switch instance.Status.Phase {
+	case kubevirtv1.Succeeded, kubevirtv1.Failed:
+		klog.Infof("instance %s is shut down.", instance.Name)
+		return true, nil
+	case kubevirtv1.Unknown:
+		klog.Warningf("instance %s is in an unknown state.", instance.Name)
+		return false, nil
```

Comment:
Need to return error here, because you can't know if it was shutdown or not
I think need to keep similar like it was before: https://github.com/kubevirt/cloud-provider-kubevirt/blob/c1b3b77c0647fe4a52f374fc638b2afe33e166b1/pkg/cloudprovider/kubevirt/instances.go#L227

Reply:
Right, fixed.